
# Week 2 — Predict Titanic Survival (Classification Model)

This section builds a Logistic Regression model to predict passenger survival, 
using the cleaned dataset from Week 1.

### 1. Setup & Load Data
Importing necessary machine learning libraries and loading the cleaned dataset from Week 1.

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix

# Load the CLEANED data from the data folder
df = pd.read_csv('../data/titanic_cleaned.csv')

### 2. Feature Engineering (Categorical Encoding)
Machine learning models require numerical input. Here, we convert categorical columns (`Sex` and `Embarked`) into numerical format using One-Hot Encoding.

In [3]:
df_encoded = pd.get_dummies(df, columns=['Sex', 'Embarked'], drop_first=True)
df_encoded.head()

,PassengerId,Survived,Pclass,Name,Age,SibSp,Parch,Ticket,Fare,Has_Cabin,Sex_male,Embarked_Q,Embarked_S
0,1,0,3,"Braund, Mr. Owen Harris",22.0,1,0,A/5 21171,7.2500,0,True,False,True
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,1,0,PC 17599,71.2833,1,False,False,False
2,3,1,3,"Heikkinen, Miss. Laina",26.0,0,0,STON/O2. 3101282,7.9250,0,False,False,True
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,1,0,113803,53.1000,1,False,False,True
4,5,0,3,"Allen, Mr. William Henry",35.0,0,0,373450,8.0500,0,True,False,True


### 3. Feature Selection & Data Splitting
Separating the target variable (`Survived`) from the predictive features. The dataset is then split into training (80%) and testing (20%) sets to evaluate the model on unseen data.

In [4]:
# Features (X) and target (y)
X = df_encoded.drop(['Survived', 'PassengerId', 'Name', 'Ticket'], axis=1)
y = df_encoded['Survived']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Training set size:", X_train.shape)
print("Test set size:", X_test.shape)

Training set size: (712, 9)
Test set size: (179, 9)


### 4. Train the Logistic Regression Model
Initializing the Logistic Regression algorithm and training it on our training dataset.

In [5]:
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

print("Model trained successfully")

Model trained successfully


### 5. Model Evaluation (Accuracy)
Making predictions on the unseen test set and calculating the overall accuracy of the model.

In [6]:
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)

print("Model Accuracy:", accuracy)

Model Accuracy: 0.8212290502793296


### 6. Confusion Matrix
Generating a confusion matrix to understand where the model is making correct predictions and where it is making errors (False Positives vs. False Negatives).

In [7]:
cm = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cm)

Confusion Matrix:
[[91 14]
 [18 56]]


### Confusion Matrix Explanation

|  | Predicted: Not Survived | Predicted: Survived |
|---|---|---|
| **Actual: Not Survived** | 91 (True Negative) | 14 (False Positive) |
| **Actual: Survived** | 18 (False Negative) | 56 (True Positive) |

- **91 True Negatives** — correctly predicted "did not survive"
- **56 True Positives** — correctly predicted "survived"
- **14 False Positives** — model predicted "survived" but they didn't
- **18 False Negatives** — model predicted "did not survive" but they actually did

Out of 179 test passengers, the model correctly predicted 147 (91 + 56), 
matching the 82.1% accuracy score. The model is slightly better at predicting 
non-survivors correctly than survivors, likely because more passengers in the 
dataset did not survive overall.